Implement provider based GameRegistry

modwizcode opened the pull request:
Currently the `GameRegistry` class is a monster, especially implementation wise. All of the different builder instances, factories and despite the class's name, only a few registries are all combined into a massive single class. 

This pull request aims to change that massively convoluted system with a series of modules, called `Provider`s that provide (hence the name) a subset of what `GameRegistry` did. The system isn't meant to be used by plugins, so all Providers are expected to be available upon request.

I've implemented the changes in a branch of SpongeCommon and was able to successfully start up a server after making numerous changes to replace hard-coded registry calls.

I know this is a drastic change, in some ways more drastic than the Data API simply because of what it encompasses, but it's needed to prevent future cruft from the over packed class.

ST-DDT commented:
Why do you have a Factory and a Provider for everything?
IMO The factory is an implementation detail, how to implement the provider.

You have also tried to split everything based on type builder, factory, statistic, career.
But it would be very convenient, if i could get a `StatisticBuilder` from the `StatisticProvider`.
And a list of all `Career`s from the `CareerProvider`.

Why do you still need the `GameRegistry` after this?

There are also some places where you replaced the text "factory" with "provider" in classes that explicitly call themselves a factory.
https://github.com/SpongePowered/SpongeAPI/pull/717/files#diff-b5be1bf622309d8523a3e70772fabaf8L67

modwizcode commented:
That was fixes in a later commit. I also don't have any class or interface
called `Factory`

ST-DDT commented:
> I also don't have any class or interface called `Factory`

[`src/main/java/org/spongepowered/api/util/factory/Factory.java`
](https://github.com/SpongePowered/SpongeAPI/pull/717/files#diff-eb0747589682ef08cc03f5838e7c493bR30) or am I missing something here?

modwizcode commented:
Sorry that's a commit error I'll have to fix that. Those were deleted.
On Jun 7, 2015 5:50 AM, "ST-DDT" <notifications@github.com> wrote:

> I also don't have any class or interface called Factory
>
> src/main/java/org/spongepowered/api/util/factory/Factory.java
>
> <https://github.com/SpongePowered/SpongeAPI/pull/717/files#diff-eb0747589682ef08cc03f5838e7c493bR30>
> or am I missing something here?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/SpongePowered/SpongeAPI/pull/717#issuecomment-109731367>
> .
>

modwizcode commented:
GameRegistry is still a good class for things that actually need a generic registry classes. There are a few things that don't fit into the provider model which will remain in `GameRegistry`

ST-DDT commented on src/main/java/org/spongepowered/api/GameRegistry.java:
```diff
@@ -25,57 +25,13 @@
-import org.spongepowered.api.block.BlockType;
-import org.spongepowered.api.data.DataManipulatorRegistry;
-import org.spongepowered.api.data.ImmutableDataRegistry;
 import org.spongepowered.api.data.type.Career;
```
Some imports survived although they are no longer required.

kashike commented on src/main/java/org/spongepowered/api/resourcepack/ResourcePackFactory.java:
```diff
@@ -28,7 +28,7 @@
 
 /**
- * A factory for creating {@link ResourcePack}s.
+ * A provider for creating {@link ResourcePack}s.
```
I have a feeling your find & replace borked this, along with other event-related classes.

modwizcode replied:
Just a bit. I'll go through and revise javadoc after the core content and
concept are accepted or rejected.

modwizcode commented:
I have made the fixes locally and will push later tonight.

gratimax commented:
You probably want to rename TextFactory to TextProvider as well. Nice work so far :+1:

modwizcode commented:
Thanks @gratimax, what are the changes to TextFactory you're interested in? It wasn't already in GameRegistry iirc so I didn't change it. I was going to change some existing classes to meet the Provider system, but it seemed too messy as most providers were in one package and a few are scattered throughout the API, which just seemed like an unwise decision.

gratimax commented:
I pretty much am only interested in the rename which is in the scope of this PR. It's not really a factory and the name "provider" just fits better. I don't think we have much use of giving the instance to plugin authors, but maybe @Minecrell has an opinion here.

simon816 commented on src/main/java/org/spongepowered/api/Game.java:
```diff
@@ -75,6 +77,15 @@
+     * @param <T> The type of {@link Provider}
+     * @return An instance of the requested provider type
+     */
+    <T extends Provider> T getProvider(Class<T> providerClass);
```
Why is this here __and__ in `GameRegistry`? Should only be in one place.

modwizcode replied:
Because technically a provider is contained in the registry, but as it was
already convoluted to get the game registry and then retrieve the instance.

By using providers it makes more sense to get them from the Game instance.
However it still made some sense to leave them with the registry itself. I
think I'll remove it. I meant to ask about opinions for the inclusion there
but I forgot.

simon816 commented on src/main/java/org/spongepowered/api/GameRegistry.java:
```diff
@@ -92,190 +48,13 @@
-     * @param <T> The type of {@link CatalogType}
-     * @return A collection of all known types of the requested catalog type
-     */
-    <T extends CatalogType> Collection<T> getAllOf(Class<T> typeClass);
```
I think this should stay because it will just make for longer method call chains and `getProvider(CatalogTypeProvider.class).getAllOf` may not be immediately obvious

modwizcode replied:
But it woudn't make any sense to put there. It's not that much longer and I think a lot of our API isn't immediately obvious which is why we use the javadocs and docs site to convey the needed information. This could easily be mentioned in the CatalogedBy annotation or other parts of the docs.

simon816 commented on src/main/java/org/spongepowered/api/GameRegistry.java:
```diff
@@ -92,190 +48,13 @@
-     * @param <T> The type of catalog type
-     * @return The found catalog type, if available
-     */
-    <T extends CatalogType> Optional<T> getType(Class<T> typeClass, String id);
```
Same again for getType as `getProvider(CatalogTypeProvider.class).getType` is less obvious

modwizcode commented:
@SpongePowered/developers I'd like to get this merged, so if people could take a look and mention changes to make I'd appreciate it. The faster this is merged the faster I can update SpongeCommon to use the new system.

Deamon5550 commented:
I"m very much against this change. You haven't really 'solved' the problem you've just moved things around and hidden it behind another layer. Perhaps if your providers were discrete to an individual part of the api (such as if you had a `StatisticsProvider` that had all the getters, builders, factories etc. relevant to statistics), but right now things like the [BuilderProvider](https://github.com/SpongePowered/SpongeAPI/pull/717/files#diff-d963bec62e2dae4434cb83b6b68f173aR47) really aren't helping the issue of "the GameRegistry class is a monster".

Additionally your `getProvider` method needs to return an optional.

modwizcode commented:
@Deamon5550 It actually doesn't need to be optional, did game registry ever return itself as optional? No, providers aren't modified or set by plugins, Sponge should always return a non-optional provider. I had them optional briefly before realizing that made no sense for an internally controlled system. A null provider is a fatal error.

I appreciate the feedback, but I don't see how I haven't "solved" the problem. Maybe it's not perfect, but it's much better than GameRegistry. Of course I've just moved things around, the GameRegistry is the one way for plugins to get instances, those methods still need to exist.

 Nothing is preventing me from moving related builders, factories and the like into their own providers, I just split it up the way the GameRegistry was organized.

Deamon5550 commented:
My biggest issue is just that your providers are all over the place. If I use statistics as an example again I get the simple statistics from the `CatalogTypeProvider`, I get some of the other statistics from `StatisticProvider `, and I get the builders for them from the `BuildersProvider`. Three separate places all while just working with statistics. If you want to make the move to providers then I think your goal for the system should be to get things into their own self contained providers so that if I'm working with statistics, all I need is the `StatisticsProvider`.

modwizcode commented:
Yeah that makes a lot of sense. I'm not exactly sure what to do with `CatalogTypeProvider` since that's used so often. It doesn't really belong in StatisticsProvider for instance. I agree builders should be fit into the relevant other providers if there are other more specific providers related to that builder.

Deamon5550 commented:
Add back a simple `Statistic get(String name);` method to the `StatisticsProvider` and still have the `CatalogTypeProvider` for more abstract fetching of data (and stuff with no provider). Although the act of having to retrieve the CatalogTypeProvider when dealing with catalog types (which you have to do fairly regularily) is rather annoying, I wonder if that is one thing which is better left out of the providers and left in GameRegistry.

modwizcode commented:
It's almost the same length, but much more clear to get the provider and
get the type as it to get the registry from the game and get the catalog
type.

On Mon, Jun 22, 2015 at 2:34 PM, Nick Conde <notifications@github.com>
wrote:

> Add back a simple Statistic get(String name); method to the
> StatisticsProvider and still have the CatalogTypeProvider for more
> abstract fetching of data (and stuff with no provider). Although the act of
> having to retrieve the CatalogTypeProvider when dealing with catalog types
> (which you have to do fairly regularily) is rather annoying, I wonder if
> that is one thing which is better left out of the providers and left in
> GameRegistry.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/SpongePowered/SpongeAPI/pull/717#issuecomment-114231173>
> .
>



-- 

Starbuck Johnson, Java Developer
www.modwiz.com | starbuck@modwiz.com

Deamon5550 commented:
In that case I think that you should remove the GameRegistry entirely, `getProvider` is in Game, `Collection<String> getDefaultGameRules()` should be in a provider, and `registerWorldGeneratorModifier(WorldGeneratorModifier modifier)` can move to Game.

modwizcode commented:
But `registerWorldGeneratorModifier(WorldGeneratorModifier modifier)` doesn't seem like it really belongs in `Game`

Deamon5550 commented:
Well otherwise its literally the only thing in GameRegistry so find a home for it because it certainly doesn't belong in GameRegistry anymore.

ST-DDT commented:
I guess we can move it out of GameRegistry to any location until we find a good place for it.

How about `Server`? There already are some `World` methods in there.

ST-DDT commented:
Two suggestions from https://github.com/SpongePowered/SpongeAPI/issues/710
* Make get `CatalogedType`s by id case insensitive (and explicitly state this in the docs)
* Add the possibility to add custom `CatalogedType`s

Zidane commented:
This needs to be redone (same as Maps) and @modwizcode isn't available so closing.

modwizcode commented:
?
On Sep 3, 2015 3:52 PM, "Chris Sanders" <notifications@github.com> wrote:

> Closed #717 <https://github.com/SpongePowered/SpongeAPI/pull/717>.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/SpongePowered/SpongeAPI/pull/717#event-400535440>.
>
